conda-store-server/conda_store_server/action/download_packages.py

```diff
@@ -28,14 +43,149 @@ def action_fetch_and_extract_conda_packages(
             file_path = pkgs_dir / filename
             count_message = f"{packages_searched} of {total_packages}"
             with filelock.FileLock(str(lock_filename)):
+                # This magic file, which is currently set to "urls.txt", is used
+                # to check cache permissions in conda, see _check_writable in
+                # PackageCacheData.
+                #
+                # Sometimes this file is not yet created while this action is
+                # running. Without this magic file, PackageCacheData cache
+                # query functions, like query_all, will return nothing.
+                #
+                # If the magic file is not present, this error might be thrown
+                # during the lockfile install action:
+                #
+                # File "/opt/conda/lib/python3.10/site-packages/conda/misc.py", line 110, in explicit
+                #   raise AssertionError("No package cache records found")
+                #
+                # The code below is from create_package_cache_directory in
+                # conda, which creates the package cache, but we only need the
+                # magic file part here:
+                cache_magic_file = pkgs_dir / PACKAGE_CACHE_MAGIC_FILE
+                if not cache_magic_file.exists():
+                    sudo_safe = expand(pkgs_dir).startswith(expand("~"))
+                    touch(cache_magic_file, mkdir=True, sudo_safe=sudo_safe)
+
                 if file_path.exists():
                     context.log.info(f"SKIPPING {filename} | FILE EXISTS\n")
                 else:
-                    context.log.info(f"DOWNLOAD {filename} | {count_message}\n")
-                    (
-                        filename,
-                        conda_package_stream,
-                    ) = conda_package_streaming.url.conda_reader_for_url(url)
-                    with file_path.open("wb") as f:
-                        shutil.copyfileobj(conda_package_stream, f)
-                    conda_package_handling.api.extract(str(file_path))
+                    with tempfile.TemporaryDirectory() as tmp_dir:
+                        tmp_dir = pathlib.Path(tmp_dir)
+                        file_path = tmp_dir / filename
+                        file_path_str = str(file_path)
+                        extracted_dir = pathlib.Path(
+                            strip_pkg_extension(file_path_str)[0]
+                        )
+                        extracted_dir_str = str(extracted_dir)
+                        context.log.info(f"DOWNLOAD {filename} | {count_message}\n")
+                        (
+                            filename,
+                            conda_package_stream,
+                        ) = conda_package_streaming.url.conda_reader_for_url(url)
```

Comment:
Let's add `conda-package-streaming` to our environment files since we are using it directly.